### PR TITLE
[Merged by Bors] - feat(CategoryTheory): a presheaf preserves certain products iff it `IsSheafFor` certain presieves 

### PR DESCRIPTION
**Pull request by dagurtomas**
A presheaf preserving a particular product `IsSheafFor` the corresponding `Presieve.ofArrows`. Conversely, if a presheaf `IsSheafFor` a `Presieve.ofArrows` and the empty presive on an initial object, then it preserves the corresponding product.

---

- [x] depends on: #8302 
- [x] depends on: #7918 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Sites/Preserves.lean [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+  cases' hf with i
+  exact Empty.elim i
+
+instance : IsEmpty (Σ(Y : C), {f : Y ⟶ ⊥_ C //
+    (Presieve.ofArrows (X := ⊥_ C) Empty.elim instIsEmptyEmpty.elim) f}) := by
+  constructor
+  rintro ⟨_, _, ⟨i⟩⟩
+  exact Empty.elim i
+
+lemma isoTerminalComparisonOfIsSheafForEmpty : IsIso (terminalComparison F) := by
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Sites/Preserves.lean:31]: Might be related: `Sheaf.isTerminalOfBotCover`

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Sites/Preserves.lean]: Thanks for pointing this out! It's not exactly what I need, but I could imitate that proof to get a shorter one here

**Comment by joelriou** [comment]
I am a little bit worried by the need to remove initial objects. I think the right path is to define a specialized API for testing the sheaf condition for presieves obtained by `Presieve.ofArrows`. As far as I can remember, and this is confirmed by reading one of the sources https://stacks.math.columbia.edu/tag/00VM there is no condition that the map from the index set `I` to the category of arrows is injective. Then, I think that an alternate "equalizer diagram" using the original index type `I` should be defined for these presieves, and it should not be hard to get the equivalence with `IsSheafFor`. Using this approach would certainly clarify the proofs in this PR, and this would also provide a very practical API for other users of the Grothendieck topologies.

**Comment by dagurtomas** [comment]
I see what you mean, I agree! I'll try to prove this sheaf condition for arrow presieves.

**Comment by digama0** [comment]
Note: I have pushed an update to the lean toolchain because this PR was on a [buggy version](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Potential.20data.20loss.20from.20.60lake.20clean.60.20with.204.2E2.2E0-rc2.2F3/near/397875701) of the toolchain. **WARNING**: checking out old commits of this PR using v4.2.0-rc2 or v4.2.0-rc3 can cause `lake clean` to delete your mathlib folder! If you need to do so, make sure to delete `lakefile.olean` manually before running any `lake` commands.

**Comment by dagurtomas** [comment]
> ... one of the sources https://stacks.math.columbia.edu/tag/00VM there is no condition that the map from the index set `I` to the category of arrows is injective. Then, I think that an alternate "equalizer diagram" using the original index type `I` should be defined for these presieves, and it should not be hard to get the equivalence with `IsSheafFor`. 

Let's define `IsSheafFor'` as the sheaf condition stated in the Stacks project for a particular `Presieve.ofArrows` (or rather, it's a condition on a family of morphisms with fixed target, repetitions allowed). Then that condition is in fact strictly stronger than the `IsSheafFor` condition of Mathlib. Building on an example you gave me on Zulip: consider a category with one object and one morphism. Then take the `Presieve.ofArrows` corresponding to a family of two copies of that morphism, and any presheaf sending the object to any non-terminal object in `Type`. Then this presheaf satisfies the `IsSheafFor` condition for that presieve, but not the `IsSheafFor'` condition.

This example of course only arises as a consequence of a stupid way of specifying a presieve consisting of a single arrow. But I hadn't noticed this subtlety before.

File: Mathlib/CategoryTheory/Sites/Preserves.lean [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+variable {C : Type u} [Category.{v} C] (F : Cᵒᵖ ⥤ Type (max u v)) [HasInitial C]
+    (hF : (ofArrows (X := ⊥_ C) Empty.elim instIsEmptyEmpty.elim).IsSheafFor F)
+
+/--
+If `F` is a presheaf which satisfies the sheaf condition with respect to the empty presieve on the
+initial object, then `F` takes the initial object to the terminal object.
+-/
+noncomputable
+def isTerminal_obj_initial_of_isSheafFor_empty_presieve : IsTerminal (F.obj (op (⊥_ C))) := by
```
Comment (b-mehta) [inline Mathlib/CategoryTheory/Sites/Preserves.lean:44]: This (and `hF`) should be stated in terms of the empty presieve on an initial object, and similar below

**Comment by joelriou** [comment]
> Let's define `IsSheafFor'` as the sheaf condition stated in the Stacks project for a particular `Presieve.ofArrows` (or rather, it's a condition on a family of morphisms with fixed target, repetitions allowed). Then that condition is in fact strictly stronger than the `IsSheafFor` condition of Mathlib. Building on an example you gave me on Zulip: consider a category with one object and one morphism. Then take the `Presieve.ofArrows` corresponding to a family of two copies of that morphism, and any presheaf sending the object to any non-terminal object in `Type`. Then this presheaf satisfies the `IsSheafFor` condition for that presieve, but not the `IsSheafFor'` condition.

Are you 100% sure it is a counterexample? It seems to me it is not. (The question is whether this non-terminal object `X`
in `Type` is the equalizer of two particular maps `X^2 ⟶ X^4`, and I believe it is.)

> This example of course only arises as a consequence of a stupid way of specifying a presieve consisting of a single arrow. But I hadn't noticed this subtlety before.

Families with duplicates are not pathological. Just consider a topological space `X`, and two open subsets `U` and `V`, then if `W` is the union of `U` and `V`, we may certainly consider the family `(U ⟶ W, V ⟶ W)`. Then, if `F` is a presheaf, whatever it is, the sheaf condition should say that `F(W)` is the pullback of `F(U)` and `F(V)` over `F(U intersection V)`. I hope this remains true even if by accident `U = V`!

(It seems to me that the notion of presieve is only an implementation detail which may have been necessary for the definition of pretopologies. Otherwise, the API should be more developed for sieves generated by a family of arrows.)

**Comment by dagurtomas** [comment]
Ok I have both directions now at #7855. I'll clean it up tomorrow, add the pullback case and some more API.

**Comment by dagurtomas** [comment]
The first part of this new API (#7855) is awaiting CI and then ready for review. It's the sheaf condition phrased in terms of unique amalgamations. I plan to add the "categorical one" in a new PR.

**Comment by joelriou** [comment]
I have left some comments on the dependent PR https://github.com/leanprover-community/mathlib4/pull/7949 which actually apply to this PR.

**Comment by dagurtomas** [comment]
I've updated both PRs

File: Mathlib/CategoryTheory/Sites/Preserves.lean [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+variable {α : Type} (X : α → C) [HasCoproduct X] [(ofArrows X (Sigma.ι X)).hasPullbacks]
+    (hd : ∀ i j [HasInitial C], i ≠ j →
+      IsPullback (initial.to _) (initial.to _) (Sigma.ι X i) (Sigma.ι X j))
+    [∀ i, Mono (Sigma.ι X i)]
+
+/--
+The two parallel maps in the equalizer diagram for the sheaf condition corresponding to the
+inclusion maps in a disjoint coproduct are equal.
+-/
+theorem firstMap_eq_secondMap : Equalizer.Presieve.Arrows.firstMap F X (fun j ↦ Sigma.ι X j) =
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/Preserves.lean:84]:
```suggestion
theorem firstMap_eq_secondMap : Equalizer.Presieve.Arrows.firstMap F X (Sigma.ι X) =
```

File: Mathlib/CategoryTheory/Sites/Preserves.lean [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    let i₂ : op (⊥_ C) ≅ (⊤_ Cᵒᵖ) :=
+      (terminalIsoIsTerminal (terminalOpOfInitial initialIsInitial)).symm
+    apply_fun (F.mapIso i₁ ≪≫ F.mapIso i₂ ≪≫ (PreservesTerminal.iso F)).hom using
+      injective_of_mono _
+    simp
+
+theorem piComparison_fac' {Z : C} (π : (i : α) → X i ⟶ Z) [IsIso (Sigma.desc π)] :
+    piComparison F (fun x ↦ op (X x)) =
+    F.map ((opCoproductIsoProduct X).inv ≫ (inv (Sigma.desc π)).op) ≫
+    Equalizer.Presieve.Arrows.forkMap F X π := by
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/Preserves.lean:108]:
```suggestion
    piComparison F (fun x ↦ op (X x)) =
      F.map ((opCoproductIsoProduct X).inv ≫ (inv (Sigma.desc π)).op) ≫
        Equalizer.Presieve.Arrows.forkMap F X π := by
```
I am not sure we specifically have a code style rule for this, but I would prefer this. As you like!

File: Mathlib/CategoryTheory/Sites/Preserves.lean [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+      inv ((Sigma.desc π).op ≫ (opCoproductIsoProduct X).hom) := by
+    simp only [op_inv, IsIso.inv_comp, IsIso.Iso.inv_hom]
+  simp only [Equalizer.Presieve.Arrows.forkMap, h₂, h₁,  ← desc_op_comp_opCoproductIsoProduct_hom,
+    ← Category.assoc, ← Functor.map_comp, IsIso.inv_comp, IsIso.Iso.inv_hom,
+    Category.assoc, IsIso.inv_hom_id, Category.comp_id, Iso.inv_hom_id, Functor.map_id,
+    Category.id_comp]
+
+theorem piComparison_fac : piComparison F (fun x ↦ op (X x)) =
+    F.map (opCoproductIsoProduct X).inv ≫
+    Equalizer.Presieve.Arrows.forkMap F X (Sigma.ι X) := by
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/Preserves.lean:121]:
```suggestion
      Equalizer.Presieve.Arrows.forkMap F X (Sigma.ι X) := by
```

File: Mathlib/CategoryTheory/Sites/Preserves.lean [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+section Product
+
+variable (hI : IsInitial I)
+
+-- This is the data of a particular disjoint coproduct in `C`.
+variable {α : Type} (X : α → C) [HasCoproduct X] [(ofArrows X (Sigma.ι X)).hasPullbacks]
+    (hd : ∀ i j [HasInitial C], i ≠ j →
+      IsPullback (initial.to _) (initial.to _) (Sigma.ι X i) (Sigma.ι X j))
+    [∀ i, Mono (Sigma.ι X i)]
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/Preserves.lean:78]: I think that a more mathlib phrasing of the definitions and lemmas in this section would be to take as input `c : Cofan X` that is colimit. The `HasInitial` should be put out of the `hd` assumption. Then, `Sigma.ι` would be replaced by `c.inj`.

**Review by joelriou — CHANGES_REQUESTED** [review]

File: Mathlib/CategoryTheory/Sites/Preserves.lean (hunk 6 above)
Comment (dagurtomas) [reply Mathlib/CategoryTheory/Sites/Preserves.lean:78]: I'll make this change later tonight or tomorrow.

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Sites/Preserves.lean:78]: I started doing this, and `firstMap_eq_secondMap` is fine with the general cocone, but in the rest of the file I'm relying heavily on the particular choice of product/coproduct by using `piComparison`. So introducing this general cocone would mean that the approach would have to change a lot for the rest of the file.

Comment (joelriou) [reply Mathlib/CategoryTheory/Sites/Preserves.lean]: I am afraid this is the right approach here :-(

Comment (joelriou) [reply Mathlib/CategoryTheory/Sites/Preserves.lean]: I do not mind if the internals of the proof uses the `piComparison`, but the main lemmas like `isSheafFor_of_preservesProduct` and ` isSheafFor_iff_preservesProduct` should be more general. `isSheafFor_of_preservesProduct` is already quite close to what I would like to get: the data `π : (i : α) → X i ⟶ Z` is mostly that of a `Cofan`, and the assumption `[IsIso (Sigma.desc π)]` is equivalent to saying that the cofan is colimit. Then, for this lemma, in the worst case scenario, if we change the assumptions as I suggest, the proof *could* start by the proof that `IsIso (Sigma.desc π)` holds, and the rest of the proof should not have to change much.

File: Mathlib/CategoryTheory/Limits/Shapes/Products.lean [hunk 7]
```diff
@@ -254,6 +254,26 @@ abbrev Sigma.desc {f : β → C} [HasCoproduct f] {P : C} (p : ∀ b, f b ⟶ P)
...
+    simp only [Cofan.mk_pt, cofan_mk_inj, this, Category.assoc]
+    simp only [Cofan.mk_pt, IsIso.hom_inv_id_assoc, colimit.ι_desc, Cofan.mk_ι_app]
+    rfl
+  · intro t m h
+    simp only [Cofan.mk_pt, IsIso.eq_inv_comp]
+    ext j
+    simp only [colimit.ι_desc_assoc, Discrete.functor_obj, Cofan.mk_pt, Cofan.mk_ι_app,
+      colimit.ι_desc]
+    exact h j
+
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Limits/Shapes/Products.lean:276]: The proof can be much shortened using the existing limits API:
```suggestion
@[simps!]
def Cofan.ext {f : β → C} {c₁ c₂ : Cofan f} (e : c₁.pt ≅ c₂.pt)
    (w : ∀ (b : β), c₁.inj b ≫ e.hom = c₂.inj b := by aesop_cat) : c₁ ≅ c₂ :=
  Cocones.ext e (fun ⟨j⟩ => w j)
  
def Cofan.isColimitOfIsIsoSigmaDesc {f : β → C} [HasCoproduct f] (c : Cofan f)
    [hc : IsIso (Sigma.desc c.inj)] : IsColimit c :=
  IsColimit.ofIsoColimit (colimit.isColimit (Discrete.functor f))
    (Cofan.ext (@asIso _ _ _ _ _ hc) (fun _ => colimit.ι_desc _ _))
```

If you absolutely need the version with `Cofan.mk`, you may add one more definition def 
```lean
def Cofan.isColimitMkOfIsIsoSigmaDesc {f : β → C} [HasCoproduct f] {P : C} (p : ∀ b, f b ⟶ P)
    [hp : IsIso (Sigma.desc p)] : IsColimit (Cofan.mk P p) := sorry
```

File: Mathlib/CategoryTheory/Limits/Opposites.lean [hunk 8]
```diff
@@ -414,44 +414,64 @@ instance : HasProduct (fun z ↦ op (Z z)) := hasLimitOfIso
...
 def opCoproductIsoProduct : op (∐ Z) ≅ ∏ (fun z => op (Z z)) :=
   IsLimit.conePointUniqueUpToIso (isLimitCoconeOp _ (coproductIsCoproduct fun b ↦ Z b))
     (limit.isLimit _) ≪≫ (IsLimit.conePointsIsoOfEquivalence
     (productIsProduct (fun z ↦ op (Z z))) (limit.isLimit _) (Discrete.opposite α).symm
     (Discrete.natIsoFunctor ≪≫ Discrete.natIso (fun _ ↦ by rfl))).symm
 
-lemma opCoproductIsoProduct_inv_comp_ι (b : α) :
-    (opCoproductIsoProduct Z).inv ≫ (Sigma.ι (fun a => Z a) b).op =
+lemma opCoproductIsoProduct_inv_comp_ι' {c : Cofan Z} (hc : IsColimit c) (b : α) :
+    (opCoproductIsoProduct Z).inv ≫
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Limits/Opposites.lean:425]: This is a good effort! But I think we should push this one last step further by refactoring the definition of `opCoproductIsoProduct` as `opCoproductIsoProduct'` which would take as inputs a colimit cofan and a limit fan fan in the opposite category, like:
```lean

variable {C : Type*} [Category C] {α : Type*} {Z : α → C}

def Cofan.op {c : Cofan Z} : Fan (fun a ↦ op (Z a)) := Fan.mk _ (fun a => (c.inj a).op)

def Cofan.IsColimit.op {c : Cofan Z} (hc : IsColimit c) : IsLimit c.op := by
  let e : Discrete.functor (fun a => Opposite.op (Z a)) ≅ (Discrete.opposite α).inverse ⋙ (Discrete.functor Z).op :=
    Discrete.natIso (fun _ => Iso.refl _)
  refine' IsLimit.ofIsoLimit ((IsLimit.postcomposeInvEquiv e _).2
    (IsLimit.whiskerEquivalence hc.op (Discrete.opposite α).symm))
    (Cones.ext (Iso.refl _) (fun ⟨a⟩ => _))
  dsimp
  erw [Category.id_comp, Category.comp_id]
  rfl

def opCoproductIsoProduct' {c : Cofan Z} {f : Fan (fun a ↦ op (Z a))}
    (hc : IsColimit c) (hf : IsLimit f) : op c.pt ≅ f.pt :=
  IsLimit.conePointUniqueUpToIso (Cofan.IsColimit.op hc) hf

def opCoproductIsoProduct :
    op (∐ Z) ≅ ∏ (fun z => op (Z z)) :=
  opCoproductIsoProduct' (coproductIsCoproduct Z) (productIsProduct (fun z ↦ op (Z z)))
```
This should also simplify the proofs of equational lemmas for `opCoproductIsoProduct` and `opCoproductIsoProduct'`.

Comment (joelriou) [reply Mathlib/CategoryTheory/Limits/Opposites.lean:425]: Note that it is probably better to do this in a separate PR.

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Limits/Opposites.lean]: @joelriou #8302 (I did the dual as well)

File: Mathlib/CategoryTheory/Sites/Preserves.lean [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+variable (hF : (ofArrows (X := I) Empty.elim instIsEmptyEmpty.elim).IsSheafFor F)
+
+section Terminal
+
+variable (I) in
+/--
+If `F` is a presheaf which satisfies the sheaf condition with respect to the empty presieve on any
+object, then `F` takes that object to the terminal object.
+If `F` is a presheaf which satisfies the sheaf condition with respect to the empty presieve on any
+object, then `F` takes that object to the terminal object.
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/Preserves.lean:50]:
```suggestion
```

File: Mathlib/CategoryTheory/Sites/Preserves.lean [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+cone.
+-/
+theorem isSheafFor_of_preservesProduct' {Z : C} (π : (i : α) → X i ⟶ Z) [HasCoproduct X]
+    [IsIso (Sigma.desc π)] [PreservesLimit (Discrete.functor (fun x ↦ op (X x))) F]
+    [(ofArrows X π).hasPullbacks] : (ofArrows X π).IsSheafFor F :=
+  have : (ofArrows X (Cofan.mk Z π).inj).hasPullbacks :=
+    (inferInstance : (ofArrows X π).hasPullbacks)
+  have : IsIso (Sigma.desc (Cofan.mk Z π).inj) := (inferInstance : IsIso (Sigma.desc π))
+  isSheafFor_of_preservesProduct F (Cofan.mk Z π) (Cofan.mk Z π).isColimitOfIsIsoSigmaDesc
+
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/Preserves.lean:160]: Do you plan to use this lemma in future PR? If not, we may remove it.
```suggestion
```

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Sites/Preserves.lean:160]: I use it in #7949

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Sites/Preserves.lean:158]: I'll probably need to refactor that PR a bit anyway once this is merged, and maybe I can just use the un-primed version of this theorem directly. Let's remove it for now and I'll add it if I really need it.

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**